package baidu.galaxy;

//////////////////////////////////////////
////////user
//////////////////////////////////////////
message User {
    optional string user = 1;
    optional string token = 2;
}

message Quota {
    optional int64 millicore = 1;
    optional int64 memory = 2;
    optional int64 storage = 3;
    optional int64 pods = 4;
}

enum Authority {
    kAuthoritySubmit = 1;
    kAuthorityRemove = 2;
    kAuthorityUpdate = 3;
    kAuthorityRead = 4;
}

enum AuthorityAction {
    kActionAdd = 1;
    kActionRemove = 2;
    kActionSet = 3;
    kActionClear = 4;
}

message Grant {
    optional User user = 1;
    repeated string tags = 2;
    optional AuthorityAction action = 3;
    repeated Authority authority = 4;
}


///////////////////////////////////////////
///// resource
///////////////////////////////////////////
message CpuResource {
    optional int64 total = 1;
    optional int64 assigned = 2;
    optional int64 used = 3;
}

message MemoryResource {
    optional int64 total = 1;
    optional int64 assigned = 2;
    optional int64 used = 3;
}

enum VolumMedium {
    KSsd = 1;
    KDisk = 2;
    KBfs = 3;
    KTmpfs = 4;
}

message Volum {
    optional string path = 1;
    optional int64 size = 2;
    optional VolumMedium medium = 3;
}

message VolumResource {
    repeated Volum total = 1;
    repeated Volum assigned = 2;
    repeated Volum used = 3;
}

message CpuRequired {
    optional int64 milli_core = 1;
    optional bool excess = 2;
}

message MemoryRequired {
    optional int64 size = 1;
    optional bool excess = 2;
}

message PortRequired {
    optional uint32 port_size = 1;
    optional uint32 start_port = 2;
}

enum VolumType {
    KEmptyDir = 1;  
    KHostDir = 2;
}

message VolumRequired {
    optional int64 size = 1;
    optional VolumType type = 2;
    optional VolumMedium medium = 3;
    optional string source_path = 4;  // bind source_path --> dest_path
    optional string dest_path = 5;    
    optional bool readonly = 6;      // default: false
    optional bool exclusive = 7;     // default: false
}

/////////////////////////////////////////
///// Job
/////////////////////////////////////////
enum JobType {
    kJobBatch = 1;
    kJobService = 2;
    kJobMonitor = 3;
}

// update incrementally may cause new problems: version confusion 
// Appmaster must save descrition of last 2 versions, and  a complete desc.
enum JobStatus {
   kJobPending = 1;
   kJobDepolying = 2;
   kJobRunning = 3;
   //kUpdating = 4;
   kJobFinished = 5;
}

// Task or pod
enum TaskStatus {
    kTaskDeploying = 1;   
    kTaskRunning = 2;
    kTaskFailed = 3;
    kTaskFinished = 4;
}

message Package {
    optional string source_path = 1; // support nfs/hdfs/ftp
    optional string dest_path = 2;   // local path
}
message ExePackage {
    optional Package package = 1;
    optional string pre_cmd = 2;     // executed befor overwriting the old package, eg: stop service
    optional string post_cmd = 3;    // execute after overwriting the old package, eg: start service
}

message DataPackage {
    repeated Package packages = 1;
    optional string pre_cmd = 2;     // 
    optional string post_cmd = 3;    // 
}

message Deploy {
    optional uint32 replica = 1;
    optional uint32 step = 2;
    optional uint32 interval = 3;
    optional uint32 max_per_host = 4;
}

message Service {
    optional string name = 1;  // mast be uniq, which is guaranteed by user
    optional uint32 port = 2;
    optional bool use_bns = 3;
}

message TaskId {
    optional string id = 1;
}

message TaskDescription {
    // set by user
    optional TaskId id = 1;

    // resource
    optional CpuRequired cpu = 2;
    optional MemoryRequired memory = 3;
    optional PortRequired ports = 4;

    // package
    optional ExePackage exe_package = 5;
    optional DataPackage data_package = 6;

    // service
    repeated Service services = 7;
}

message JobId {
    optional string id = 1;
}

// user -> resource manager
message PodDescription {
    // Job
    optional string name = 1;
    optional JobType type = 2;

    // workspace volum
    optional VolumRequired workspace_volum = 3;

    // Resource, volum is owned by Pod
    repeated VolumRequired data_volums = 4;
    
    // tasks 
    repeated TaskDescription tasks = 5;

    // schedule-policy is necessary ?
}

message JobDescription {
    optional Deploy deploy = 1;
    optional PodDescription job_desc = 2;
}


///////////////////////////////////////
////container
///////////////////////////////////////

message CgroupId {
    optional string id = 1; 
}

message Cgroup {
   optional CgroupId id = 1; 
   optional CpuRequired cpu = 2;
   optional MemoryRequired memory = 3;
    
}

message ContainerGroup {
    optional string id = 1;
}

message ContainerId {
    optional string id = 1;     //
}

// resource manager -> agent
message ContainerDescription {
   // optional ContainerGroup = 1; 
   // optional ContainerId id = 2;
    optional VolumRequired workspace_volum = 3;
    repeated VolumRequired data_volums = 4;      //
    optional string cmd_line = 5;                // appmaster or appwork
    repeated Cgroup cgroups = 6;
}


enum ContainerStatus {
    kContainerPending = 1;    // in pending queue
    kContainerIniting = 2;    // initing
    kContainerReady = 3;      // ready, no process in container 
    kContainerRunning = 4;    // running 
    kContainerError = 5;      // an error occure
    kContainerFinished = 6;   // released by resource manager
}

// agent -> manager
message ContainerInfo {
    optional string id = 1; 
    optional int64 created_time = 2;
    optional ContainerStatus status = 3; 

    // resource?
    
    // statistics, eg: error cnt or other exceptions
}

///////////////////////////////////////
////status
///////////////////////////////////////
enum Status {
   kOk = 1; 
}

message ErrorCode {
    optional Status status = 1;
    optional string reason = 2;
}

// agent -> resource manager
message AgentInfo {
    // agent version

    // host healthy

    // container
    repeated ContainerInfo container_info = 1;

    // resource
    optional CpuResource cpu_resoruce = 2;
    optional MemoryResource memory_resource = 3;
    repeated VolumResource volum_resources = 4;

    // exception statistics, eg: failed num of pod ..
}
